job-server/src/main/scala/spark/jobserver/util/JsonProtocols.scala

```diff
@@ -1,16 +1,17 @@
 package spark.jobserver.util
 
 import akka.http.scaladsl.model.Uri
-import org.joda.time.DateTime
 import spark.jobserver.common.akka.web.JsonUtils.AnyJsonFormat
 
-import java.text.SimpleDateFormat
 import spark.jobserver.io.{BinaryInfo, BinaryType, ContextInfo, JobInfo, JobStatus}
 import spray.json._
 
+import java.time.format.DateTimeFormatter
+import java.time.{ZoneId, ZonedDateTime}
+
 object JsonProtocols extends DefaultJsonProtocol {
 
-  val DATE_PATTERN = "yyyy-MM-dd HH-mm-ss SS Z"
+  val DATE_PATTERN = "yyyy-MM-dd'T'HH:mm:ss.SSSX"
```

Comment:
Will this change break reading existing data from the database?
If I already have some contexts/jobs data in DB, will they be parsed correctly by the new Jobserver?
There is no such test in this change.

Reply:
Ok, two things:
1. The pattern was actually wrong, I have corrected it and added two test cases to check it. The expected strings are extracted from the Readme but I also verified that the old behaviour of SJS also conforms to this pattern.
2. I am not sure where this particular pattern was actually used. I was not able to find it in any user-exposed data types/interfaces. They all use the new pattern (which is identical to the default of JodaTime.toString()).

Reply:
This JSON protocols are at least used in Zookeeper DAO.
I now made a simple test by starting a Zookeeper docker container and configuring DAO as described in https://github.com/spark-jobserver/spark-jobserver/blob/master/doc/dao-setup.md#hdfs--zookeeper

I created a context and could see it in Zookeeper and in CLI:
```
[zk: localhost:2181(CONNECTED) 13] get /jobserver/db/contexts/94a9e08a-4edc-4843-b377-e4eb639b278e
{"actorAddress":null,"config":"{\"context\":{\"name\":\"test\"},\"context-factory\":\"spark.jobserver.context.DefaultSparkContextFactory\",\"context-init-timeout\":\"60s\",\"forked-jvm-init-timeout\":\"30s\",\"is-adhoc\":\"false\",\"launcher\":{\"spark\":{\"driver\":{\"memory\":\"1g\"}}},\"memory-per-node\":\"512m\",\"num-cpu-cores\":2,\"passthrough\":{\"spark\":{\"driver\":{\"allowMultipleContexts\":true}}},\"python\":{\"executable\":\"python3\",\"paths\":[]},\"spark\":{\"session\":{\"hive\":{\"enabled\":true}}},\"streaming\":{\"batch_interval\":1000,\"stopGracefully\":true,\"stopSparkContext\":true}}","endTime":"2021-05-11 11-56-24 531 +0200","error":"Context failed to connect back within initialization time","id":"94a9e08a-4edc-4843-b377-e4eb639b278e","name":"test","startTime":"2021-05-11 11-56-24 532 +0200","state":"ERROR"}
```
```
curl localhost:8090/contexts/test
{"name":"test","startTime":"2021-05-11T11:56:24.532+02:00","state":"ERROR","endTime":"2021-05-11T11:56:24.531+02:00","id":"94a9e08a-4edc-4843-b377-e4eb639b278e"}
```


After starting Jobserver with current change, context metadata was not anymore readable:
```
[2021-05-11 12:01:48,911] ERROR o.zookeeper.ZookeeperUtils [] [akka://JobServer/user/dao-manager] - Text '2021-05-11 11-56-24 531 +0200' could not be parsed
at index 10 : java.time.format.DateTimeParseException: Text '2021-05-11 11-56-24 531 +0200' could not be parsed at index 10
        at java.time.format.DateTimeFormatter.parseResolved0(DateTimeFormatter.java:1949) 
```

I didn't do such test for SQL DAO yet, but I think it also should be checked.